rust/protocol/src/sealed_sender.rs

```diff
@@ -757,37 +757,56 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
     Ok(serialized)
 }
 
+/// This is a single-key multi-recipient KEM, defined in Manuel Barbosa's
+/// ["Randomness Reuse: Extensions and Improvements"]. It uses the "Generic Construction" in `4.1`
+/// of that paper, instantiated with ElGamal encryption.
+///
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// ### Overview
+/// 1. generate a random [sealed_sender_v2::DerivedKeys].
+/// 2. encrypt the keypair using a shared secret derived from the sender's private key and
+///    recipient's public key (this secret never changes for that pair of identity keys) with
+///    [sealed_sender_v2::apply_agreement_xor].
+/// 3. sign the encrypted keypair using the sender's private key with
+///    [sealed_sender_v2::compute_authentication_tag].
+/// 4. use the keypair from (1) to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent].
```

Comment:
This is a little oversimplified. Maybe it's
1. Generate a series of random bytes.
2. Derive an ephemeral key pair from (1).
3. Encrypt (1) using a shared secret derived from the private ephemeral key (2) and the recipient's public identity key…
4. Add an authentication tag for (3) using a secret derived from the sender's private identity key and the recipient's public identity key.
5. Use the key pair from (2) to…
6. Send the public ephemeral key (2), the encrypted random bytes (3), and the authentication tag (4) to the recipient along with the encrypted message (5).

But also while I think you've done a nice job summarizing what the algorithm does, this is all still implementation detail docs (it's on a private module, after all), and we already have the crypto-paper-style pseudocode (not my personal favorite, but conventional!). What made you decide to add a prose high-level summary of the algorithm? Clearly you felt it was worth it or you wouldn't have bothered to write it out!

Reply:
So I'm totally fine leaving it the way it is if you'd prefer to keep the private implementation private, as this PR was largely an excuse to hash out my understanding of this process with you! My other reasoning for making this PR was to alleviate my confusion on which parts of this process were handled by the Signal backend server, since in my fork I have extended this sealed-sender v2 mechanism to use when sending encrypted `PreKeyBundle`s (see https://github.com/cosmicexplorer/libsignal-client/commit/a20a58e5aa85d3d6571f4f4e915ff2ef2c5e459a for a very quick summary of the extensions). I was a bit flummoxed at first by what actually went into sealed-sender messages after 18463e8357ef6c8489568b0c0489c2161740842a, since the messages sent over the wire (reasonably!) no longer corresponded directly to a protobuf schema.

Context: I am orchestrating this "encrypted `PreKeyBundle`" mechanism from within an (also AGPL) crate (not public yet) that adapts this codebase to be used in places where the `gpg` tool is normally used, i.e. for sending encrypted messages across insecure channels. For my next trick I would like to also extend the sealed-sender concept to encrypt the message recipient as well, since my impression is that unlike the Signal app which requires the recipient address for routing (mentioned in passing in [this blog post](https://signal.org/blog/sealed-sender/)), my tool will be used in a context where I can ensure messages get routed securely without leaking metadata (even if this requires additional overhead vs the Signal app). If I succeed in doing this, I would actually like to propose pushing these changes upstream, as I see them as enabling additional metadata resistance for Signal conversations, but I wanted to spend the time to develop these techniques first in the context of my own project before doing so.

So to reiterate, the addition of this rather intense docstring is motivated by my own application of the Signal client codebase, which has explicitly different goals than the mainstream Signal app. I accept that upstream will likely change these sorts of implementation details in the future and have been eagerly rebasing against them. I do *not* want to slow down the Signal app by [federating it and making it beholden to lots of external projects that require a stable API](https://signal.org/blog/the-ecosystem-is-moving/). So, feel free to propose closing this PR if you're uncomfortable with that.